### PR TITLE
Refactor data processing

### LINKED ISSUE #7: work on data processing steps for swarm positioning
Spend a short amount of time looking for obvious ways to speed up the data processing steps for that chart. This will make developing those charts much faster.

Notes:
- time intensive to find positions in swarm. currently takes 18 min to process the smallest chunk - 2% for 20 years
- chunking by time needs to be done thoughtfully to ensure that events crossing time breaks are included in full duration
- `create_event_swarm` fxn may be worth exploring

### PR DESCRIPTION
**Pull request by hcorson-dosch**
This PR closes #7 

This work has been undertaken in a series of steps, somewhat iteratively. I did my best to commit the code associated with each step separately, but some changes were inadvertently lumped.

1. Per the latest meeting with the drought team, use only the variable 7d drought properties
2. Add in simplest approach to date chunking
3. Speed up `create_event_swarm()` function w/o altering its approach (avoid full refactoring)
4. Make a minor edit to the approach used in `create_event_swarm()`  and `event_swarm_plot` to speed up plotting

What I've avoiding undertaking is a full refactoring of the function, but it might be necessary down the road if we eventually wanted to speed up data processing for the higher threshold values (10%+). I have some notes about what I think could work, but haven't explored it.

-----------------------------
# Steps 1 and 2
The work for step 1 and 2 is primarily in [this commit](https://github.com/USGS-VIZLAB/drought-timeline/commit/6fc953535b43aa8a45e99f71974c8f28edb263e4) and [this commit](https://github.com/USGS-VIZLAB/drought-timeline/commit/24f5a263ef842f3c04a565bf56c00352cefff397)

These steps were focused on addressing the second bullet in #7 ("chunking by time needs to be done thoughtfully to ensure that events crossing time breaks are included in full duration")

**TLDR: Split the data on dates where there are no droughts, then process in chunks**

_Longer description (to have as a record of these changes):_
For any given set of drought properties data, `identify_drought_chunks()` uses dates without drought to divide the event period into different temporal 'chunks', with an option to set the size of the chunks. For each chunk we compute a few statistics, one key one being the maximum number of droughts on any single day within each chunk.

Once the chunks have been identified, we can map over those chunks, calling `create_event_swarm()` [for each chunk](https://github.com/hcorson-dosch/drought-timeline/blob/refactor_data_processing/2_process.R#L54-L60). For each chunk, we can use the maximum number of droughts on any single day within each chunk to [set the number of rows in the event matrix in `create_event_swarm()`](https://github.com/hcorson-dosch/drought-timeline/blob/refactor_data_processing/2_process/src/prep_stripswarm.R#L28) (forgot to include that snippet in ^ commits). Previously an arbitrary value was used, which made the matrix unnecessarily large.

This code helped a lot, at least in the lower percentile ranges where gaps were present, as it allowed the data processing to be done in chunks. But the function still ran slowly, particularly with longer stretches (~5+ years).

---------------------------
# Step 3

The work for step 3 is in [this commit](https://github.com/USGS-VIZLAB/drought-timeline/commit/06ed80437a5d088da2dffff3e5423bc2384faf2d) and [this commit](https://github.com/USGS-VIZLAB/drought-timeline/commit/ca38c06f455fbb9113a5215e4253f994c3780130)

The next step was focused on making the low-hanging changes to speed up `create_event_swarm()` without changing the approach of the function or fully refactoring it.

**TLDR: speed up `create_event_swarm` by dropping expensive sampling and using `data.table`**

_Longer description (to have as a record of these changes):_
For this step, I took a look at the `create_event_swarm()` function with `profvis`. I made 2 main sets of changes:
  - `Profvis` identified that the [slice_sample() call](https://github.com/USGS-VIZLAB/drought-timeline/blob/main/2_process/src/prep_stripswarm.R#L38), was slow. Since the function ultimately [always pulled the first row from the lowest priority group](https://github.com/USGS-VIZLAB/drought-timeline/blob/main/2_process/src/prep_stripswarm.R#L40-L41), I switched to filtering to row(s) with the minimum priority value _before_ sampling.
  - The other functions that `profvis` identified were dplyr ones - `select`, `group_by`, etc. so I re-wrote the looped portion of the function with `data.table`.
  
These changes significantly sped up the data processing, but the code still runs slowly when processing long periods with lots of droughts where chunking isn't possible (e.g., 10% threshold at the national scale for 1951-2020)

As an additional benefit, these changes fixed a bug with the original function. The # of rows of the dataframe returned by `create_event_swarm` should equal the # of drought days (unique site days with drought). The original `create_event_swarm` returned a different value. The function re-written with data.table returns a dataframe with the correct number of rows.

-----------------------------
# Step 4

The work for step 4 is in [this commit](https://github.com/USGS-VIZLAB/drought-timeline/commit/4b38d014606d9471ae019dd5ed92b4e98126981a) and [this commit](https://github.com/USGS-VIZLAB/drought-timeline/commit/9064d83949b979dc03fc9708ad3d7abfa204b732)

This step was focused on making a quick change to speed up the plotting.

**TLDR: return one row per drought event instead of one row per drought day - use duration to set width of tile**

_Longer description (to have as a record of these changes):_
While iterating on figure versions (#6 ) I noticed that for periods with lots of drought events, the plotting was really struggling, b/c the combined dataframe returned by the chunked calls to `create_event_swarm()` was huge. The existing approach returned a dataframe with a row for every drought day (unique site day with drought). When plotted, each row became a single square tile. I modified the approach to return a single row for each drought event, and then used the stored `duration` value to set the width of the tile when plotting.

This change really sped up plotting for those periods with more events. With the changes in step 3, the data at a national level, 2% threshold, from 1951-2020 could process in ~25 minutes instead of 1hr+. But then plotting took 2 hours. With the new functions, the data processing is the same, but the plotting took 25 seconds (!!).

I've left `create_event_swarm_compressed()` and `event_swarm_plot_compressed()` as separate functions for now b/c I wanted to get your take, Cee, on whether these produce sufficiently identical results to go with the faster approach. If so, I'll add an additional commit that replaces the original functions with these new ones.

Here's an example (note that the placement of drought events in rows has a random element in each approach):

Original approach:
swarm_jd7d_2_2010_2019

'Compressed' approach:
swarm_jd7d_2_2010_2019_compressed

---------------------------
# Other edits

As an aside, I also [updated the CASC region defintions](https://github.com/USGS-VIZLAB/drought-timeline/commit/5818799bb16cb5ab7c0ace8d93715439faaaa350), as 2 states were missing

---------------------------
# Unaddressed issues

As noted in a few places above, the changes in steps 2 and 3 have sped up processing, but the gains are primarily when working with shorter periods or with lower drought thresholds. I think improving processing time for the higher thresholds (where there are more drought events and more overlapping drought events) would take some additional refactoring. One option would be to further modify the chunking approach to allow for the user to set break points, but then there would need to be a mechanism for accounting for and correctly added drought events that span those break points. I have some ideas, but haven't explored anything as it didn't seem worth the investment at the moment.

### CODE REVIEW

File: 2_process.R [hunk 1]
```diff
@@ -8,10 +8,11 @@ p2_targets <- list(
...
              # Climate Adaptation Regions
              list(NW = c('WA','OR','ID'),
-                  SW = c('CA','UT','NV','NM'),
+                  SW = c('CA','UT','NV','AZ'),
+                  SC = c('OK','TX','LA','NM'),
                   NC = c('MT','ND','SD','WY','CO','NE','KS'),
-                  MW = c('WI','MN','IA','IN','IL','OH','MI'),
-                  NE = c('ME','VT','NH','NY','NJ','MA','RI','CT','WV','VA','MD','DE','KY'),
+                  MW = c('WI','MN','IA','IN','IL','OH','MI', 'MO'),
+                  NE = c('ME','VT','NH','NY','NJ','PA','MA','RI','CT','WV','VA','MD','DE','KY'),
```
Comment (ceenell) [inline 2_process.R:15]: 😥

File: 1_fetch.R [hunk 2]
```diff
@@ -11,21 +11,12 @@ p1_targets <- list(
...
-                      sb_secret_exists = p0_sb_credentials_exist)
-  },
-  format = 'file'),
+  ###### Download drought properties file ######
   tar_target(p1_1951_2020_drought_prop_jd_7d_csv, {
     # Depend on this dummy variable to initiate re-download of files
     p0_sb_fetch_date
     download_sb_files(sb_id = p0_sbitem_child_1951_2020,
-                      sb_files_to_download = 'Drought_Properties_jd_07d_wndw_updated.csv',
+                      sb_files_to_download = 'Drought_Properties_jd_07d_wndw.csv',
```
Comment (ceenell) [inline 1_fetch.R:19]: Thanks for updating this. I think `study_watersheds_metadata.csv` also needs to be subbed for the `gagesII` metadata file?

Comment (hcorson-dosch) [reply 1_fetch.R:19]: Thanks for catching that! I should have double check it. Just pushed a commit that makes that change

**Review by ceenell — APPROVED** [review]
Really nice job with this. Everything built for me as expected. You made some smart choices at each step outlined here that have made it much more efficient without having to invest in a complete rework. A+

Changing the plotting also gives us a little more flexibility. I like that you’ve left both option here though for now, and think we should leave it like that for the time being. I say that because depending on how we want the output to be structured. 

One thing in the back of my mind is that if we wanted to create navigation to certain time periods, like decades, it could be really straightforward to chunk the output by decade and navigate to the parent div. In that case, we’d want to keep the original, so the edge between decades would align but be separate. This could also be used to drive lazy loading. The compressed may work equally as well, with navigation elements overlaid on the big png and using the compressed version. I think some nav exploration is necessary to move forward here.

Let’s forget about the other thresholds for now and proceed as if we will just use the lowest. We can reassess this decision if there is strong feedback on it, but it seems like we have a solid direction here to move out on that doesn’t require additional refactoring.

**Comment by hcorson-dosch** [comment]
> One thing in the back of my mind is that if we wanted to create navigation to certain time periods, like decades, it could be really straightforward to chunk the output by decade and navigate to the parent div. In that case, we’d want to keep the original, so the edge between decades would align but be separate. This could also be used to drive lazy loading. The compressed may work equally as well, with navigation elements overlaid on the big png and using the compressed version. I think some nav exploration is necessary to move forward here.

Yes! I had the same thought when I was first making that compressed approach, but then totally forgot. So glad you added that note here. Thank you!

As a side note, I realized you can get the uncompressed version of the swarm from the compressed like so:
```r
revert_compressed <- p2_nat_2_swarm_compressed  %>% 
  uncount(duration, .remove=FALSE, .id="id") %>%
  mutate(start_date = date, date = date+id-1, .after=id)
```

That sounds good to me to keep the 2 approaches for now, and to move forward with the lowest threshold. 

Thanks for the review, Cee!
